document/core/syntax/types.rst

```diff
@@ -31,15 +31,72 @@ Integers are not inherently signed or unsigned, their interpretation is determin
 The types |F32| and |F64| classify 32 and 64 bit floating-point data, respectively.
 They correspond to the respective binary floating-point representations, also known as *single* and *double* precision, as defined by the |IEEE754|_ standard (Section 3.3).
 
+Number types are *transparent*, meaning that their bit patterns can be observed.
+Values of number type can be stored in :ref:`memories <syntax-mem>`.
+
 Conventions
 ...........
 
-* The meta variable :math:`t` ranges over value types where clear from context.
-
-* The notation :math:`|t|` denotes the *bit width* of a value type.
+* The notation :math:`|t|` denotes the *bit width* of a number type :math:`t`.
   That is, :math:`|\I32| = |\F32| = 32` and :math:`|\I64| = |\F64| = 64`.
 
 
+.. index:: ! reference type, reference, table, function, function type, null
+   pair: abstract syntax; reference type
+   pair: reference; type
+.. _syntax-reftype:
+
+Reference Types
+~~~~~~~~~~~~~~~
+
+*Reference types* classify first-class references to objects in the runtime :ref:`store <store>`.
+
+.. math::
+   \begin{array}{llll}
+   \production{reference type} & \reftype &::=&
+     \ANYREF ~|~ \ANYFUNC ~|~ \EQREF ~|~ \NULLREF \\
+   \end{array}
+
+The type |ANYREF| denotes the infinite union of all references, and thereby a :ref:`supertype <match-reftype>` of all other reference types.
+
+The type |ANYFUNC| denotes the infinite union of all references to :ref:`functions <syntax-func>`, regardless of their :ref:`function types <syntax-functype>`.
+
+The type |EQREF| denotes the infinite union of all references that can be compared for equality;
+in order to avoid exposing implementation details, some reference types, such as |ANYFUNC|, do not admit equality, and therefore are not :ref:`subtypes <match-reftype>` of |EQREF|.
+
+The type |NULLREF| only contains a single value: the :ref:`null <syntax-ref_null>` reference.
+It is a :ref:`subtype <match-reftype>` of all other reference types.
+The |NULLREF| type cannot be used in a program, it only occurs during :ref:`validation <valid>`.
```

Comment:
I'm trying to see how this is the case: as is, it appears that `nullref` is a `valtype` which means it can appear in function signatures and as the type of locals.  So you can have silly expressions like:
```
(local $x nulltype)
(call $foo (tee_local $x (get_local $x)))
```
Instead of trying to introduce a validation-only type, can't we just have an operator `reftype.null` where `reftype` is later constrained by validation to only allow nullable types?

Reply:
Ah hah, I hadn't noticed `nullref` wasn't part of concrete syntax!  Yes, that does solve the problem rather elegantly in the spec.  In the implementation, though, what will still be annoying is that we'll probably need separate valtype enums for what can appear in locals/signatures (the latter of which takes part in serialization and indirect call ABI matching so we need to be crystal clear about what the cases are) and for what can appear as part of validation.  And we'll have to go back and forth between them and ask "which valtype enum do I use here?" and it's not a huge deal, but it'd be nice to avoid if there's not a downside.

> And it seems a bit wasteful to have separate null instructions for each type constructor

For this I was assuming a single bytecode "null" followed by the [reftype](https://webassembly.github.io/reference-types/core/binary/types.html#reference-types) as an immediate (which handles both the primitive and compound types in the encoding)...

> We could type-annotate the instruction instead, but that also just seems unnecessary given that we're already buying into subtyping anyway.

... which is what I think you mean here.  Agreed we are necessarily buying into subtyping, the only difference here is that the other types will be present in the concrete syntax and so there's not really another option (other than, I suppose, removing subtyping in lieu of a static `upcast` operator).  With null, it seems like we have an option and the non-subtype route seems mildly simpler.

Reply:
> This seems like a point worth discussing at the April meeting.

Yeah, happy to discuss in the group.  In the short-term, we were wanting to get going on a prototype impl of this proposal and, because of aforementioned valtype warts, we'd prefer to implement the type-parameterized `null`.  Could you perhaps flip the proposal to that so we can directly implement the proposal (knowing we might have to change if the proposal flips back)?

Reply:
Well, it's a few hours work on tests and interpreter, too, that I probably want to postpone until it's clear that I won't have to undo it shortly after. But I'm happy to hold back landing these PRs until the meeting. I don’t think that needs to block you from prototyping either version.